src/gui/transferlistwidget.cpp

```diff
@@ -878,17 +954,24 @@ void TransferListWidget::displayListMenu(const QPoint&)
     QList<QAction *> tagsActions;
     QMenu *tagsMenu = listMenu.addMenu(GuiIconProvider::instance()->getIcon("view-categories"), tr("Tags"));
     tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("list-add"), tr("Add...", "Add / assign multiple tags..."));
-    tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("edit-clear"), tr("Remove...", "Remove multiple tags..."));
     tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("edit-clear"), tr("Remove All", "Remove all tags"));
     tagsMenu->addSeparator();
     foreach (QString tag, tags) {
-        const bool setChecked = tagsInSelection.contains(tag);
-        tag.replace('&', "&&");  // avoid '&' becomes accelerator key
-        QAction *tagSelection = new QAction(GuiIconProvider::instance()->getIcon("inode-directory"), tag, tagsMenu);
-        tagSelection->setCheckable(true);
-        tagSelection->setChecked(setChecked);
-        tagsMenu->addAction(tagSelection);
-        tagsActions << tagSelection;
+        const Qt::CheckState initialState = tagsInAll.contains(tag) ? Qt::Checked
+                                            : tagsInAny.contains(tag) ? Qt::PartiallyChecked
+                                            : Qt::Unchecked;
+
+        const CheckBoxMenuItem::ToggleFn onToggle = [this, tag](Qt::CheckState newState)
+        {
+            if (newState == Qt::CheckState::Checked)
+                addSelectionTag(tag);
+            else if (newState == Qt::CheckState::Unchecked)
+                removeSelectionTag(tag);
+            else
+                qWarning() << "Unexpected checkstate: " << newState;
```

Comment:
It would indicate a bug.

There is a third state, Qt::CheckState::PartiallyChecked, but it should never be the case that we are toggling to that state. Checkboxes that do not start in the PartiallyChecked state have that state disabled. For boxes that do start in PartiallyChecked, we disable the state as soon as the user toggles away from it to any other state.

We also do a static_cast from int to CheckState when connecting this method to the stateChanged signal from the checkbox. That int is supposed to map to CheckState, but if there were a bug in stateChanged and it passed an invalid value, that would show up here.